File: examples/leetcode-out/hs/5/longest-palindromic-substring.hs
module Main where

import Data.Maybe (fromMaybe)


forLoop :: Int -> Int -> (Int -> Maybe a) -> Maybe a
forLoop start end f = go start
  where
    go i | i < end =
            case f i of
              Just v -> Just v
              Nothing -> go (i + 1)
         | otherwise = Nothing

avg :: Real a => [a] -> Double
avg xs | null xs = 0
      | otherwise = sum (map realToFrac xs) / fromIntegral (length xs)

_indexString :: String -> Int -> String
_indexString s i =
  let idx = if i < 0 then i + length s else i
  in if idx < 0 || idx >= length s
       then error "index out of range"
       else [s !! idx]

_input :: IO String
_input = getLine


expand s left right = fromMaybe (0) $
    (let l = left in (let r = right in (let n = length s in Just (((r - l) - 1)))))

longestPalindrome s = fromMaybe ("") $
    case if (length s <= 1) then Just (s) else Nothing of Just v -> Just v; Nothing -> (let start = 0 in (let end = 0 in (let n = length s in case forLoop 0 n (\i -> (let len1 = expand s i i in (let len2 = expand s i (i + 1) in (let l = len1 in case if (len2 > len1) then (let l = len2 in Nothing) else Nothing of Just v -> Just v; Nothing -> if ((l > end) - start) then (let start = (div (i - ((l - 1))) 2) in (let end = (div (i + l) 2) in Nothing)) else Nothing)))) of Just v -> Just v; Nothing -> Just ((s !! start)))))

main :: IO ()
main = do
    return ()

File: examples/leetcode-out/pas/5/longest-palindromic-substring.pas
program main;
{$mode objfpc}
uses SysUtils, fgl;

type
	generic TArray<T> = array of T;

function expand(s: string; left: integer; right: integer): integer;
var
	l: integer;
	n: integer;
	r: integer;
begin
	l := left;
	r := right;
	n := Length(s);
	while ((l >= 0) and (r < n)) do
	begin
		if (s[l + 1] <> s[r + 1]) then
		begin
			break;
		end;
		l := l - 1;
		r := r + 1;
	end;
	result := r - l - 1;
	exit;
end;

function longestPalindrome(s: string): string;
var
	_end: integer;
	i: integer;
	l: integer;
	len1: integer;
	len2: integer;
	n: integer;
	start: integer;
begin
	if (Length(s) <= 1) then
	begin
		result := s;
		exit;
	end;
	start := 0;
	_end := 0;
	n := Length(s);
	for i := 0 to n - 1 do
	begin
		len1 := expand(s, i, i);
		len2 := expand(s, i, i + 1);
		l := len1;
		if (len2 > len1) then
		begin
			l := len2;
		end;
		if (l > _end - start) then
		begin
			start := i - l - 1 div 2;
			_end := i + l div 2;
		end;
	end;
	result := Copy(s, start + 1, (_end + 1 - start));
	exit;
end;

begin
end.
